### DISCUSSION TITLE
Best Alternative to ytdl-core and youtube-dl

### DISCUSSION THREAD

**Discussion by iim-ayush** [discordjs/voice#188] [General]
I have made [play-dl](https://github.com/play-dl/play-dl) to overcome 2 things :

- ytdl-core stream abort issues and miniget errors
- youtube-dl launching takes a lot of time.

Play-dl gets WebM opus streams and you never need ffmpeg for normal videos. But for LiveStreams, it requires ffmpeg to convert it into opus.

IDK how, but is very fast as compared of other youtube alternatives to stream audio on discord bot

**Comment by YuzuTheNeko** [comment]
it works fine for songs that are under 10m long or so, thats because of the watermark you gave it (which is higher than ytdl one, thats why u dont get any error as often)
![image](https://user-images.githubusercontent.com/57841770/130368203-346aa2e3-d511-418a-8aad-795a5546d054.png)
Connection reset after 15-18m of the same song

**Reply by iim-ayush** [reply]
Thanks for using play-dl

First thing to clarify is that ytdl-core by default uses [10 MB](https://github.com/fent/node-ytdl-core/blob/40d0c54da10a12397ac63b0dd39ed27de4ff294b/lib/index.js#L102) as highwaterMark. So you can't compare my package with ytdl-core in respect of highWaterMark. 

The main reason why it doesn't gets error as often because I use [got](https://github.com/sindresorhus/got) where as ytdl-core uses [miniget](https://github.com/fent/node-miniget). I use my own based streaming rather than sticking to [m3u8stream](https://github.com/fent/node-m3u8stream). 

For the errors, it would be awesome if you report issue in main repository since I am going to handle this in next release.
In my mind, I thought it would create issue after long time. But I didn't thought it would give errors after 15-20 mins mark.

I will fix this in next release. It is a new package so expect some errors.

**Reply by iim-ayush** [reply]
> it works fine for songs that are under 10m long or so, thats because of the watermark you gave it (which is higher than ytdl one, thats why u dont get any error as often)
> ![image](https://user-images.githubusercontent.com/57841770/130368203-346aa2e3-d511-418a-8aad-795a5546d054.png)
> Connection reset after 15-18m of the same song

Your problem has been fixed in v0.5.0

Go check it out

**Reply by JairusSW** [reply]
play-dl is awesome! 👍🏼 
I'm going for a lot of users and concurrent streams, so I'm building and hand-optimizing my own.

**Comment by iim-ayush** [comment]
Released [v0.5.0](https://github.com/play-dl/play-dl) .

Best youtube stream creator for djs-music bots

**Reply by YuzuTheNeko** [reply]
looks good now 👍

**Comment by Mistelteyn** [comment]
ytdl-core stream abort issues really frustrated me.
so far I used 3 music that always aborted on ytdl-core but after using yours it worked fine.

Thank you!

**Comment by JairusSW** [comment]
I'm currently working on a youtube downloader that uses https://npmjs.com/package/undici as the http client. It it around 10-15x faster than miniget/got in my experience and uses quite a bit less memory. I'll push to github soon, so anyone can fork and make a PR. Its going to be awesome for bots because it can handle so much more concurrent streams compared to ytdl

**Reply by iim-ayush** [reply]
Is that a public repository ??

I can also contribute to that. Since I have a lot of experience in this area.

It would be great if you give link of the repository.

I think play-dl v1.10 upcoming one will sort many issues and will be the best in this segment.

You will encounter many problems in your way of creating a perfect YouTube downloader, so I would recommend you to join play-dl support server and ask help if you need any.

**Reply by JairusSW** [reply]
@killer069, what is the invite link to the discord?

**Reply by iim-ayush** [reply]
@JairusSW 

That should be in ReadMe at play-dl github

OR use [this](https://discord.gg/Zf8Utqde)

**Comment by JairusSW** [comment]
Cool! I'll be posting it probably in a couple of hours. Its going to be primarily on performance and concurrency. That'll be awesome if you could contribute! :)
